{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FINDING, ACCESSING, INTEGRATING, and REUSING data (FAIR)\n",
    "\n",
    "During this course, we will explore the history of bioinformatics data publication and reuse.  We will explore data technologies from the most primitive through the past ~20 years to the most contemporary.  \n",
    "\n",
    "The overview of our lessons for the next ~2 months is:\n",
    "\n",
    "* \"Screen Scraping\" data from pages directly accessed on the Web\n",
    "* Application Programming Interfaces (APIs) for data retrieval and integration\n",
    "    * Custom APIs\n",
    "    * \"REST\" APIs\n",
    "    * BioRuby APIs\n",
    "* Linked Data & SPARQL query\n",
    "* Web Publishing - static and dynamic database access  (you will NOT be tested on this!)\n",
    "* <strike>FAIR Data</strike>\n",
    "\n",
    "<strike>\n",
    "In the end, you will be required to show that you can publish data that is FAIR for both humans and machines - you will be making a Website that can be queried by both people and by computers (this is your final assignment of the year)</strike>\n",
    "<pre>\n",
    "\n",
    "\n",
    "</pre>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# The Painful Truth\n",
    "\n",
    "\n",
    "Getting data from the Web is still far more painful than it should be.\n",
    "\n",
    "\n",
    "Most resources do not offer data in a format that is easily accessible by machines.\n",
    "\n",
    "\n",
    "Many resources claim to provide “RESTful” access, but even they often make it difficult (we will discuss REST in a future lecture).\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# One example of a \"friendly\" provider\n",
    "\n",
    "There are (too) many examples of Bioinformatics data providers that do not make it easy to retrieve their data.  In your future careers, you will probably need to access data from one of these providers.   Good luck!\n",
    "\n",
    "In this course, we are going to mainly focus on \"nice\" data providers.  Nevertheless, we begin with a ~difficult case, so that you learn how to deal with these situations.  After that, and over the next weeks, we will explore increasingly better, friendlier, more powerful, and easier-to-explore data sources.\n",
    "\n",
    "For today, we are going to explore the European Bioinformatics Institute (EBI) dbFetch interface (**NOTE: I am NOT saying that this is a BAD interface!  I am just using it as an example of a difficult case from the perspective of software!**).\n",
    "\n",
    "http://www.ebi.ac.uk/Tools/dbfetch/\n",
    "\n",
    "dbFetch provides data in a variety of formats, using a predictable URL structure.\n",
    "\n",
    "<pre>\n",
    "\n",
    "\n",
    "</pre>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# From Within a Script\n",
    "\n",
    "Scroll down the dbFetch web page until you get to the section titled \"from within a script\":\n",
    "\n",
    "------------------\n",
    "\n",
    "* from within a script - examples of the URL for all styles and formats\n",
    "\n",
    "For people interested in programmatic access to the Dbfetch functionality, we recommend using our new Web Services version of Dbfetch: WSDbfetch.\n",
    "\n",
    "Alternatively, you can use dbfetch for direct access:\n",
    "\n",
    "Making scripted http requests to dbfetch is very simple, the parameters which can be used are db, id, format and style. Of these parameters only db and id are required fields. When omitting to use format and/or style, the defaults for the chosen database will be used (the default style is always html).\n",
    "\n",
    "The URL to dbfetch is always of this format:\n",
    "**dbfetch?db=DB_NAME&id=IDS&format=FORMAT_NAME&style=STYLE_NAME**\n",
    "\n",
    "    DB_NAME - Must be chosen from the table below\n",
    "    IDS - Single id/acc or comma/white-space separated list (id1 or id1 id2 id3 or id1,id2,id3)\n",
    "    FORMAT_NAME - Name of the output format, varies between databases\n",
    "    STYLE_NAME - Name of the output style, available styles are raw and html\n",
    "\n",
    "\n",
    "For details of the available databases, formats and styles see the list of databases [http://www.ebi.ac.uk/Tools/dbfetch/dbfetch/dbfetch.databases]. Additional examples are provided in the syntax guide [http://www.ebi.ac.uk/Tools/dbfetch/syntax.jsp].\n",
    "\n",
    "Examples:\n",
    "\n",
    "https://www.ebi.ac.uk/Tools/dbfetch/dbfetch?db=ena_sequence&id=J00231,K00650,D87894,AJ242600\n",
    "\n",
    "Instead of the default raw (plain ASCII) style, entries can also be retrieved in plain text (raw):\n",
    "https://www.ebi.ac.uk/Tools/dbfetch/dbfetch?db=ena_sequence&id=J00231,K00650,D87894,AJ242600&style=raw\n",
    "\n",
    "It is also possible to retrieve Fasta formatted sequences:\n",
    "https://www.ebi.ac.uk/Tools/dbfetch/dbfetch?db=ena_sequence&id=J00231,K00650,D87894,AJ242600&format=fasta\n",
    "\n",
    "Because of backward compatibility issues the program can be simply called by giving one or more INSDC accession numbers or entry names:\n",
    "https://www.ebi.ac.uk/Tools/dbfetch/dbfetch?J00231\n",
    "\n",
    "-------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# The dbFetch URL Structure\n",
    "\n",
    "## dbfetch?db=<span style='color:red;'>DB_NAME</span>&id=IDS&format=FORMAT_NAME&style=STYLE_NAME\n",
    "\n",
    "The first element to notice in the dbFetch URL is the \"DB_NAME\" field.  The value of this field can be one of many of the databaases hosted at EBI.  The list is below.   I have highlighted \"Ensembl Genomes Gene\" because it is one of the few databases that understands *Arabodiopsis* gene IDs\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# The List of Databases\n",
    "\n",
    "\n",
    "An overview of each database is also provided, which includes a short description of the database, a link to the database, a collection of example identifiers and details of the available data formats and result styles.\n",
    "Databases\n",
    "\n",
    "* EDAM (edam)\n",
    "* ENA Coding (ena_coding)\n",
    "* ENA Geospatial (ena_geospatial)\n",
    "* ENA Non-coding (ena_noncoding)\n",
    "* ENA Sequence (ena_sequence)\n",
    "* ENA Sequence Constructed (ena_sequence_con)\n",
    "* ENA Sequence Constructed Expanded (ena_sequence_conexp)\n",
    "* ENA/SVA (ena_sva)\n",
    "* Ensembl Gene (ensemblgene)\n",
    "* **Ensembl Genomes Gene (ensemblgenomesgene)**[ http://www.ebi.ac.uk/Tools/dbfetch/dbfetch/dbfetch.databases#ensemblgenomesgene ]\n",
    "* Ensembl Genomes Transcript (ensemblgenomestranscript)\n",
    "* Ensembl Transcript (ensembltranscript)\n",
    "* EPO Proteins (epo_prt)\n",
    "* HGNC (hgnc)\n",
    "* IMGT/HLA (imgthla)\n",
    "* IMGT/LIGM-DB (imgtligm)\n",
    "* InterPro (interpro)\n",
    "* IPD-KIR (ipdkir)\n",
    "* IPD-MHC (ipdmhc)\n",
    "* IPRMC (iprmc)\n",
    "* IPRMC UniParc (iprmcuniparc)\n",
    "* JPO Proteins (jpo_prt)\n",
    "* KIPO Proteins (kipo_prt)\n",
    "* MEDLINE (medline)\n",
    "* Patent DNA NRL1 (nrnl1)\n",
    "* Patent DNA NRL2 (nrnl2)\n",
    "* Patent Protein NRL1 (nrpl1)\n",
    "* Patent Protein NRL2 (nrpl2)\n",
    "* Patent Equivalents (patent_equivalents)\n",
    "* PDB (pdb)\n",
    "* RefSeq (nucleotide) (refseqn)\n",
    "* RefSeq (protein) (refseqp)\n",
    "* SGT (sgt)\n",
    "* Taxonomy (taxonomy)\n",
    "* Trace Archive (tracearchive)\n",
    "* UniParc (uniparc)\n",
    "* UniProtKB (uniprotkb)\n",
    "* UniRef100 (uniref100)\n",
    "* UniRef50 (uniref50)\n",
    "* UniRef90 (uniref90)\n",
    "* UniSave (unisave)\n",
    "* USPTO Proteins (uspto_prt)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ensembl Genomes Gene (ensemblgenomesgene)\n",
    "\n",
    "http://www.ensemblgenomes.org/\n",
    "\n",
    "Ensembl Genomes genome databases for metazoa, plants, fungi, protists and bacteria, for vertebrate species and **model organisms** [1] see Ensembl instead of Ensembl Genomes. Gene sequences and annotations.\n",
    "\n",
    "<span style='font-size: 50%;'>[1]  Apparently, EnsEMBL doesn't consider Arabidopsis a \"model organism\" ;-)</span>\n",
    "\n",
    "| <span style='color: blue;'>Format</span> |\t<span style='color: green;'>Styles</span> | \tExample Identifiers | \n",
    "| ------ | ------- | -------\n",
    "| default  | \tdefault, raw \t | Id: AAEL000001, AGAP006864, GB46163, b2736\n",
    "| csv  | \tdefault, raw  | \tId: AAEL000001, AGAP006864, GB46163, b2736\n",
    "| **embl**  | \tdefault, raw  | \tId: AAEL000001, AGAP006864, GB46163, b2736\n",
    "| fasta  | \tdefault, raw  | \tId: AAEL000001, AGAP006864, GB46163, b2736\n",
    "| genbank  | \tdefault, raw  | \tId: AAEL000001, AGAP006864, GB46163, b2736\n",
    "| gff2  | \tdefault, raw  | \tId: AAEL000001, AGAP006864, GB46163, b2736\n",
    "| gff3  | \tdefault, raw  | \tId: AAEL000001, AGAP006864, GB46163, b2736\n",
    "| tab |  \tdefault, raw  | \tId: AAEL000001, AGAP006864, GB46163, b2736 \n",
    "\n",
    "\n",
    "\n",
    "## dbfetch?db=DB_NAME&id=IDS&format=<span style='color: blue;'>FORMAT_NAME</span>&style=<span style='color: green;'>STYLE_NAME</span>\n",
    "\n",
    "The columns for Format and Style are highlighted, together with their position in our structured URL.  I have also highlighted **\"embl\"**, because that is the data format that I want us to use for this example.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# So far we have...\n",
    "\n",
    "## dbfetch?db=<span style='color: red;'>ensemblgenomesgene</span>&format=<span style='color: blue;'>embl</span>&id=.....\n",
    "\n",
    "These elements are called \"GET String Parameters\".  **db** is one parameter, with a value of \"ensemblgenomesgene\".  **format** is another parameter, with the value \"embl\".  In a URL, Get String parameters have one of these two patterns:\n",
    "\n",
    "    ?parameter1=value1;parameter2=value2;.....\n",
    "    ?parameter1=value1&parameter2=value2&.....\n",
    "   \n",
    "i.e., at the end of teh URL, the parameter/value pairs follow a '?' symbol, and parameter/value pairs are separated by '&' or ';'. (We will discuss what \"GET\" means when we talk about REST in a future lecture.)\n",
    "\n",
    "What is the complete URL?  What is before the '?'?\n",
    "\n",
    "Go back to the main dbFetch page and look at the examples:\n",
    "\n",
    "Examples:\n",
    "\n",
    "**https://www.ebi.ac.uk/Tools/dbfetch/dbfetch**?db=ena_sequence&id=J00231,K00650,D87894,AJ242600\n",
    "\n",
    "Instead of the default raw (plain ASCII) style, entries can also be retrieved in plain text (raw):\n",
    "\n",
    "**https://www.ebi.ac.uk/Tools/dbfetch/dbfetch**?db=ena_sequence&id=J00231,K00650,D87894,AJ242600&style=raw\n",
    "\n",
    "It is also possible to retrieve Fasta formatted sequences:\n",
    "\n",
    "**https://www.ebi.ac.uk/Tools/dbfetch/dbfetch**?db=ena_sequence&id=J00231,K00650,D87894,AJ242600&format=fasta\n",
    "<pre>\n",
    "\n",
    "\n",
    "</pre>\n",
    "\n",
    "\n",
    "\n",
    "##  So, the URL for the Arabidopsis gene At3g54340, in ensembl format, is:\n",
    "\n",
    "### https://www.ebi.ac.uk/Tools/dbfetch/dbfetch?db=ensemblgenomesgene&format=embl&id=At3g54340\n",
    "\n",
    "Open this in your browser (open in a new window!) and examine the structure of that file.  How would you extract information from that file, for example, the name of the gene, or the number of exons?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<pre>\n",
    "  \n",
    "\n",
    "\n",
    "\n",
    "</pre>\n",
    "  \n",
    "# Prove that you understand:\n",
    "\n",
    "* Retrieve the EMBL record for AT4G36920 (what gene name is it?)\n",
    "* Retrieve the FASTA file for AT4G36920\n",
    "* Retrieve the UniProt record for AP3_ARATH \n",
    "    * DO NOT retrieve it in HTML - i.e. you want the RECORD, not the web page containing the record --> you will need to use the “style” GET string parameter (&style=...).  Try the different \"style\" options until you get what you want.  Look at the table(s) of styles for each database to decide\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "dbfetch?db=ensemblgenomesgene&format=embl&style=raw&id=AT4G36920\n",
    "dbfetch?db=ensemblgenomesgene&format=fasta&style=raw&id=AT4G36920\n",
    "dbfetch?db=uniprotkb&format=uniprot&style=raw&id=AP3_ARATH\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<pre>\n",
    "\n",
    "\n",
    "\n",
    "</pre>\n",
    "\n",
    "# Your first step towards massively integrative informatics - Deja el navegador!\n",
    "\n",
    "It is impossible to do modern bioinformatics through your browser!  To achieve data discovery and integration at the scale of contemporary systems biology, it is necessary to do all of the data retrieval and integration inside of your software.\n",
    "\n",
    "**How do you access The Web in your code?   Use a softeware library called \"rest-client\"**\n",
    "\n",
    "\n",
    "<pre>\n",
    "\n",
    "</pre>\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "require 'rest-client'   # this is how you access the Web\n",
    "\n",
    "address = 'http://www.ebi.ac.uk/Tools/dbfetch/dbfetch?db=ensemblgenomesgene&format=embl&id=At3g54340'\n",
    "\n",
    "response = RestClient::Request.execute(\n",
    "  method: :get,\n",
    "  url: address)  # use the RestClient::Request object's method \"execute\"\n",
    "\n",
    "# puts response.methods-Object.new.methods\n",
    "\n",
    "record =  response.body\n",
    "record.match(/..../)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "<pre>\n",
    "\n",
    "\n",
    "</pre>\n",
    "\n",
    "## That was easy!  \n",
    "\n",
    "\n",
    "That was easy!   ....too easy!  This is a good time to learn about how to handle errors in Ruby.  A lot of the time when you try to access the Web, something will go wrong... maybe the Web page doesn't exist, or you have lost your WiFi connection, or you have made a mistake in your code, or... or... or...\n",
    "\n",
    "\n",
    "The code below does the same thing as the code above, but it is more careful.  It uses a function called \"fetch\" to attempt to contact the website, and if it fails, it does something useful (returns False).  This is the first time you have seen error-handling.  Look at the examples in the code, and try to understand them.  See also:  http://rubylearning.com/satishtalim/ruby_exceptions.html \n",
    "<pre>\n",
    "\n",
    "\n",
    "</pre>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "require 'rest-client'  \n",
    "\n",
    "# Create a function called \"fetch\" that we can re-use everywhere in our code\n",
    "\n",
    "def fetch(url:, headers: {accept: \"*/*\"}, user: \"\", pass: \"\")\n",
    "  response = RestClient::Request.execute({\n",
    "    method: :get,\n",
    "    url: url.to_s,\n",
    "    user: user,\n",
    "    password: pass,\n",
    "    headers: headers})\n",
    "  return response\n",
    "  \n",
    "  rescue RestClient::ExceptionWithResponse => e\n",
    "    warn e.inspect\n",
    "    response = false\n",
    "    return response  # now we are returning 'False', and we will check that with an \\\"if\\\" statement in our main code\n",
    "  rescue RestClient::Exception => e\n",
    "    $stderr.puts e.inspect\n",
    "    response = false\n",
    "    return response  # now we are returning 'False', and we will check that with an \\\"if\\\" statement in our main code\n",
    "  rescue Exception => e\n",
    "    $stderr.puts e.inspect\n",
    "    response = false\n",
    "    return response  # now we are returning 'False', and we will check that with an \\\"if\\\" statement in our main code\n",
    "end \n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "\n",
    "    \n",
    "res = fetch(url: 'http://www.ebi.ac.uk/Tools/dbfetch/dbfetch?db=ensemblgenomesgene&format=embl&id=At3g54340');\n",
    "#puts res.class\n",
    "#puts res.public_methods\n",
    "# abort  #this is how you immediately stop a Ruby programme\n",
    "\n",
    "if res  # res is either the response object (RestClient::Response), or false, so you can test it with 'if'\n",
    "  body = res.body  # get the \"body\" of the response\n",
    "  #headers = res.headers  # get other details about the HTTP message itself\n",
    "  puts body\n",
    "else\n",
    "  puts \"the Web call failed - see STDERR for details...\"\n",
    "end\n",
    "\n",
    "  \n",
    "  \n",
    "  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<pre>\n",
    "\n",
    "</pre>\n",
    "# Extracting data from the Web - \"screen scraping\"!\n",
    "\n",
    "You can now use tools that you already know - like regular expressions - to extract specific pieces of data from this Web response.  For example, the gene name:\n",
    "\n",
    "(note that this code contains more error handling.  It checks if the returned content of the web page contains a match to your regular expression, and it tries to \"rescue\" the situation if it fails to find a match.)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    \n",
    "res = fetch(url: 'http://www.ebi.ac.uk/Tools/dbfetch/dbfetch?db=ensemblgenomesgene&format=embl&id=At3g54340');\n",
    "  \n",
    "if res  # res is either the response object, or False, so you can test it with 'if'\n",
    "  body = res.body  # get the \"body\" of the response\n",
    "  \n",
    "  if body =~ /locus_tag=\"([^\"]+)\"/ \n",
    "#     gene_name = $1\n",
    "#     puts \"the name of the gene is #{gene_name}\"\n",
    "#     abort\n",
    "    \n",
    "  #puts body\n",
    "    gene_name_regexp = Regexp.new(/locus_tag=\"([^\"]+)\"/)  # this is one way to do Regular Expressions in Ruby.  There are several!\n",
    "    match = gene_name_regexp.match(body)\n",
    "    if match\n",
    "      gene_name = match[1]  # matches act like an array, so the first match is [1]\n",
    "                            # (try for yourself, what is match[0]?)\n",
    "      puts \"the name of the gene is #{gene_name}\"\n",
    "      abort\n",
    "    end\n",
    "  else\n",
    "    begin  # use a \"begin\" block to handle errors\n",
    "      puts \"There was no gene name in this record\"  # print a friendly message\n",
    "      raise \"this is an error\" # raise an exception\n",
    "    rescue # some code to rescue the situation... for example, maybe a different regexp?\n",
    "      puts \"exiting gracefully\"  # in this case, we are just going to stop trying\n",
    "    end\n",
    "  end\n",
    "end\n",
    "  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "<pre>\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "</pre>\n",
    "# Prove that you understand:\n",
    "\n",
    "### to do this, you should switch to VS Code and write the code as an independent file that uses the objects you created for your assignment - don't try to do this in Jupyter!\n",
    "\n",
    " 1.  (easy) read the gene_information.tsv file from your assignment; retrieve the EMBL record for each gene in that file  (please retrieve them one-at-a-time, not all together)\n",
    "\n",
    " 2.  (medium) Create an “AnnotatedGene” Class, that extends the Gene Class by adding the DNA sequence and the Protein sequence attributes (as Strings).\n",
    "\n",
    " 3.  (hard) For every EMBL file retrieved from gene_information.tsv, find the cross-referenced UniProt identifier (needs a regular expression). \n",
    "\n",
    "    * Retrieve the UniProt Protein record as FASTA using dbfetch.  \n",
    "    * Retrieve the DNA (FASTA) sequence using dbfetch.   \n",
    "    * Create an AnnotatedGene object for each one.  \n",
    "    * Loop over every AnnotatedGene object and create a report:\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<pre><code>\n",
    "AGI_Locus        At3GXXXXX\n",
    "GeneName            Ap3\n",
    "Protein_ID        P11234\n",
    "DNA Sequence\n",
    ">Cosa|Cosa|At3GXXXXX    (keep the original FASTA header)\n",
    "ACTGCTAGCTGATGCTGATGCATGCTAGC\n",
    "Protein Sequence\n",
    ">Cosa|Cosa|P111234  (keep the original FASTA header)\n",
    "MYLLMSSPIOO\n",
    "</code></pre>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prove that you understand by using a different Web resource\n",
    "\n",
    "__The Gene Ontology Consortium provides two kinds of files:__\n",
    "* Ontology Files (Terms, Term ID, and Descriptions)\n",
    "* Associations (The GO terms used to annotate every gene)\n",
    "\n",
    " 4. (easy) In your code, retrieve the GO Slim Plant Ontology:\n",
    "http://www.geneontology.org/ontology/subsets/goslim_plant.obo\n",
    "\n",
    " 5. (~easy) In your code, parse that file, and for a GO identifier (e.g. GO:0006950) print the GO Term name to the screen (e.g. “response to stress”) (there are MANY different solutions for this!  All of them are regular expressions...)\n",
    "\n",
    " 6. (hard) Add “GO_Annotation” attribute to your AnnotatedGene Class (array of strings), then:\n",
    " \n",
    "For every gene in the gene_information.tsv file, \n",
    "* Retrieve the UniProt Record\n",
    "* Retrieve the GO annotations (GO_NNNNNN) from the UniProt record\n",
    "* Retreive the GO Term name from UniProt, and Term definition from the goslim_plant.obo file\n",
    "* Add those annotations to your AnnotatedGene object (think about this…..)\n",
    "* create a report like:\n",
    "\n",
    "\n",
    "<code>Gene_ID: \n",
    "Term1 (def)\n",
    "Term2 (def)\n",
    "Term3 (def)\n",
    "</code>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<pre>\n",
    "  \n",
    "\n",
    "</pre>\n",
    "\n",
    "  ## These were examples of difficult cases\n",
    "  \n",
    "If you have to \"screen scrape\", you are facing a very difficult situation.  Your code will often break (when the data provider changes their Web page, for example).  You also probably have not seen EVERY possible case, so your RegularExpression will probably miss some data sometimes.  \n",
    "\n",
    "If you can find a BETTER way to get the data, you should do that!  In the next lesson we will learn about REST and about other data formats like JSON and XML, that make it easier to extract the data.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Ruby 3.0.0",
   "language": "ruby",
   "name": "ruby"
  },
  "language_info": {
   "file_extension": ".rb",
   "mimetype": "application/x-ruby",
   "name": "ruby",
   "version": "3.0.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
